### ISSUE TITLE
Custom introspection

### ISSUE THREAD

**Issue by andrew-laughlin** [IdentityModel/IdentityModel.AspNetCore.OAuth2Introspection#13]
We're using IS4 1.5.0. In an effort to support multi-tenancy, we've customized the introspection endpoint (by implementing IIntrospectionRequestValidator) to allow additional scopes to be specified in the request. This allows claims return by the introspection endpoint to be in a different (tenant)context than the original authN request. It essentially allows API resources to augment requests with different or additional tenant IDs.

To make this work we need a method to allow the introspection middleware to inspect an incoming API request, and based that inspection, specify additional parameters on the introspection request. Maybe something similiar to way token retrieval works:

`string token = Options.TokenRetriever(Context.Request);`

e.g.

`IDictionary<string, string> introspectionParameters = Options.AdditionalParameterRetriever(Context.Request);`

which gets added to the introspection request:

```
public class IntrospectionRequest
{
	public string Token { get; set; }
	public IDictionary<string, string> Parameters { get; set; }
	...
}
```
There's also the issue of caching.
I hope this makes sense. Thoughts on this approach?

Thanks,
Andrew

**Comment by leastprivilege** [comment]
Wouldn't the backchannel handler be sufficient for that?

**Comment by andrew-laughlin** [comment]
It may be part of the equation, but I couldn't see how the HttpMessageHandler would get access to the original API request. Let me add some additional detail.

The request coming into the API contains a reference token and some information about tenancy -- either in an HTTP header or on the route or some other way. The tenancy information on the request needs to be translated to scopes (or some other customization) and then added to the introspection endpoint call. My thought in the initial post above, was to provide some type of hook into the process of creating the introspection request to allow generic customization of the request.

**Comment by brockallen** [comment]
This is the consequence of Microsoft designing the back channel extensibility point as a property on the options instance, and not as a service that is injected via DI.

**Comment by andrew-laughlin** [comment]
@leastprivilege @brockallen , what would you guys recommend to accomplish this?

**Comment by leastprivilege** [comment]
I don't have the time to think about this right now - I guess the first step would be to add additional parameter functionality here

https://github.com/IdentityModel/IdentityModel2/blob/dev/src/IdentityModel/Client/IntrospectionRequest.cs

**Comment by andrew-laughlin** [comment]
I've submitted a couple PRs to facilitate this scenario. Please let me know what you think.

https://github.com/IdentityModel/IdentityModel2/pull/48
https://github.com/IdentityModel/IdentityModel.AspNetCore.OAuth2Introspection/pull/14

If the PRs look good, the remaining item will be to update the IdentityModel package reference in the OAuth2Introspection project, to make the changes available.

**Comment by andrew-laughlin** [comment]
@leastprivilege @brockallen  Thanks for pulling in the IdentityModel PR. Any thoughts on the related PR?

**Comment by leastprivilege** [comment]
We started a review - you should have been notified.

**Comment by andrew-laughlin** [comment]
ok thanks. I've updated this PR to consume the IdentityModel 2.9.1 release --> https://github.com/IdentityModel/IdentityModel.AspNetCore.OAuth2Introspection/pull/14

**Comment by leastprivilege** [comment]
We just released 3.0.0 for ASP.NET Core 2 - 

we don't have the cycles to continue supporting the old version. If you want to re-consider this feature on top of the new architecture - let me know.
